kiwi/api_helper.py

```diff
@@ -0,0 +1,82 @@
+# Copyright (c) 2021 SUSE Software Solutions Germany GmbH. All rights reserved.
+#
+# This file is part of kiwi.
+#
+# kiwi is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# kiwi is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with kiwi.  If not, see <http://www.gnu.org/licenses/>
+#
+import logging
+from functools import wraps
+from typing import Callable
+
+from kiwi.exceptions import KiwiDecommissionedApiFunctionCalled
+
+log = logging.getLogger('kiwi')
+
+
+def obsolete(decommission_at: str) -> Callable:
+    """
+    Decorator method to mark an API method as obsolete
+
+    Methods marked with this decorator are still called.
+    If the provided decommission date is reached it is
+    allowed to decorate the obsoleted method as
+    decommissioned.
+
+    Example:
+
+        .. code:: python
+
+            @obsolete(decommission_at='2025-01-28')
+            def method():
+                print('Method implementation...')
+
+    :param str decommission_at: date string
+    """
+    def obsolete_method(func: Callable) -> Callable:
+        @wraps(func)
+        def function(*args, **kwargs):
+            nonlocal decommission_at
+            log.warning(
+                f'Function {func.__name__!r} is marked obsolete '
+                f'and will be decommissioned at: {decommission_at}'
```

Comment:
Just an idea, you could use the `warnings` module. With the `warnings.simplefilter` you could filter such deprecation messages.

Reply:
The kiwi api provides a logging class to handle all types of logging module controlled messages. I have implemented the use of the logging system such that you can register the kiwi logging class. If you do that as a user you benefit from the logging work done in kiwi. If you register your own logging class you can do whatever you like, if you don't setup logging you get the default python logging behavior which includes the WARN type. The idea was to handle log messages through pythons logging facility. I do not understand how the warnings module will help if the idea was to transport everything through the logging facility ?

Reply:
> I do not understand how the warnings module will help if the idea was to transport everything through the logging facility ?

Yes, I know you can emit warning messages with the logging module too. :slightly_smiling_face: The idea is a slightly different use case.

It boils down to this: Warnings from the logging module are **targeted for the normal user**. Warnings from the warnings module are **targeted to developers**. 

Sure, you can mix both. But sometimes you want to separate these two concerns.

The warnings module can help _developers_. By default, warning messages are disabled. For simple scripts it can be enabled by starting the Python script with `python -Wd SCRIPTNAME` and you can see all (even from other modules!). It's even possible to filter specific warnings.

Don't know if this is something which is really a use case for KIWI. My text was just providing you with this information, it's up to you. :slightly_smiling_face: 

The distinction between `logging.warning` and `warnings.warn` is described in this StackOverflow post: https://stackoverflow.com/a/37979724

If you want to know how to use warnings and logging together, I'd recommend to read the little section [Integration with the warnings module](https://docs.python.org/3/library/logging.html#integration-with-the-warnings-module) in the Python documentation.

Reply:
Oh, and I forge to mention another difference between these two: warnings show up only once. Quoted from official doc:

>    Repetitions of a particular warning for the same source location are typically suppressed.

Example:

```python
>>> import warnings
>>> warnings.warn("foo")
__main__:1: UserWarning: foo
>>> warnings.warn("foo")
>>> warnings.warn("foo")
```

Reply:
ok the differentiation between logging (users) and warnings(developers) makes sense to me. I will use this concept. Thanks